src/Catalog/Helpers/Db2CatalogSqlQueries.cs

```diff
@@ -0,0 +1,122 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Data;
+using System.Data.SqlClient;
+using System.Data.SqlTypes;
+using System.Threading.Tasks;
+
+namespace NuGet.Services.Metadata.Catalog.Helpers
+{
+    /// <summary>
+    /// Utility class for all SQL queries invoked by Db2Catalog.
+    /// </summary>
+    public class Db2CatalogSqlQueries : IDb2CatalogQueries
+    {
+        private const string CursorParameterName = "Cursor";
+        private readonly Db2CatalogProjection _db2catalogProjection;
+        private readonly int _commandTimeout;
+
+        public Db2CatalogSqlQueries(
+            PackageContentUriBuilder packageContentUriBuilder,
+            int commandTimeout)
+        {
+            _db2catalogProjection = new Db2CatalogProjection(packageContentUriBuilder);
+            _commandTimeout = commandTimeout;
+        }
+
+        public async Task<IReadOnlyCollection<DateTime>> GetTimestampsAsync(
+            SqlConnection sqlConnection,
+            SqlTransaction sqlTransaction,
+            int top,
+            DateTime since,
+            string columnName)
+        {
+            var nextTimestampsQuery = $@"
+                        SELECT TOP {top} P.[{columnName}]
+                        FROM [dbo].[Packages] AS P
+                        WHERE [{columnName}] IS NOT NULL AND [{columnName}] > @{CursorParameterName}
+                        GROUP BY [{columnName}]
+                        ORDER BY [{columnName}]";
+
+            var nextTimestampsCommand = new SqlCommand(nextTimestampsQuery, sqlConnection, sqlTransaction)
+            {
+                CommandTimeout = _commandTimeout
+            };
+
+            nextTimestampsCommand.Parameters.AddWithValue(
+                CursorParameterName,
+                since < SqlDateTime.MinValue ? SqlDateTime.MinValue : since);
+
+            var timestamps = new List<DateTime>();
+            using (var timestampsReader = await nextTimestampsCommand.ExecuteReaderAsync())
+            {
+                while (await timestampsReader.ReadAsync())
+                {
+                    timestamps.Add(timestampsReader.ReadDateTime(columnName).ForceUtc());
+                }
+            }
+
+            return timestamps;
+        }
+
+        public async Task<IReadOnlyCollection<FeedPackageDetails>> GetPackageDetailsForTimestampsAsync(
+            SqlConnection sqlConnection,
+            SqlTransaction sqlTransaction,
+            string columnName,
+            IReadOnlyCollection<DateTime> timestamps)
+        {
+            var packages = new List<FeedPackageDetails>();
```

Comment:
Can this lead to OutOfMemory? Should we consider a max count of Packages retrieved?